persistenceOne/persistenceCore
# Upgrade to revert tombstoning of certain validators [Do not merge]

## 1. Overview

An upgrade that reverts that tombstoning of certain validators.

## 2. Implementation details

The upgrade loops over the list of specified validators and reverts the tombstone on each of them.
The validators are still in jail afterwards, they need to be unjailed manually.
For each account that had delegated to the validators at the moment of slashing, new coins are minted to roughly reimburse them for the slashing and the lost APR since.
Finally, the coins are sent to their respective addresses.

## 3. How to test/use

- Set up a local net with a few validators.
- Double sign with one or more of them (e.g. Copy the priv_validator_key to another validator)
- Propose an upgrade to revert the tombstoning
- Edit the code and enter the addresses of the tombstoned validators
- Build the new binary and contribute it among the validators
- After the upgrade has successfully been executed, the validators are no longer tombstoned and the relevent addresses are refunded
- The validators still sit in jail and need to be unjailed maunally

## 4. Checklist



- [ ] Does the Readme need to be updated?

## 5. Limitations (optional)

At it's current state, it was not possible to actually send coins to the affected addresses.
The coins were successfully minted but the program crashed when it tried to send the coins from the mint module to account addresses because of a nil pointer dereference  (app/cos_patch.go:81)

## 6. Future Work (optional)

The affected validators and accounts need to be queried and added to the patch so they can be reimbursed.
In addition to that, the function to send coins from the mint module to the accounts needs to be debugged.
